backend/src/main/java/com/google/sps/FindMatchQuery.java

```diff
@@ -50,30 +54,107 @@ public Match findMatch(Participant newParticipant) {
-            newParticipant.getUsername(),
-            currParticipant.getUsername(),
-            duration,
-            currentTimeMillis);
+      long secondEndTimeAvailable = currParticipant.getEndTimeAvailable();
+      if (!isCompatibleTime(duration, firstEndTimeAvailable, secondEndTimeAvailable)) {
+        System.out.println("not compatible time");
+        continue;
+      }
+
+      // Check match preference compatibility and get combined preference if compatible
+      MatchPreference secondMatchPreference = currParticipant.getMatchPreference();
+      MatchPreference combinedMatchPreference =
+          getCombinedMatchPreference(firstMatchPreference, secondMatchPreference);
+      if (combinedMatchPreference == null) {
+        System.out.println("not compatible match pref");
+        continue;
       }
+
+      // Check if combined match preference is satisfied depending on number of same fields
+      if (!isCombinedMatchPreferenceSatisfied(
+          combinedMatchPreference, newParticipant, currParticipant)) {
+        continue;
+      }
+
+      // Found a match
+      System.out.println("found a match");
+      return new Match(
+          newParticipant.getUsername(), currParticipant.getUsername(), duration, clock.millis());
     }
     // No inital match found
     return null;
   }
+
+  /** Return true if compatible endTimeAvailable considering duration and padding */
+  private boolean isCompatibleTime(
+      int duration, long firstEndTimeAvailable, long secondEndTimeAvailable) {
+    long earliestEndTimeAvailable = Math.min(firstEndTimeAvailable, secondEndTimeAvailable);
+    return (clock.millis() + TimeUnit.MINUTES.toMillis(duration + PADDING_MINUTES))
+        < earliestEndTimeAvailable;
+  }
+
+  /**
+   * @return combined MatchPreference similar = both are similar OR one is similar and one is any;
+   *     any = both are any; different = both are different = one is different and one is any. or
+   *     null if match preferences are not compatible
+   */
+  private MatchPreference getCompatibleMatchPreference(
+      MatchPreference firstMatchPreference, MatchPreference secondMatchPreference) {
+    if ((firstMatchPreference == secondMatchPreference)
+        || (firstMatchPreference == MatchPreference.ANY)
+        || (secondMatchPreference == MatchPreference.ANY)) {
+      return (firstMatchPreference == MatchPreference.ANY)
+          ? secondMatchPreference
+          : firstMatchPreference;
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * @return combined MatchPreference similar: both are similar, one is similar and one is any any:
+   *     both are any different: both are different, one is different and one is any
+   */
+  private MatchPreference getCombinedMatchPreference(
+      MatchPreference firstMatchPreference, MatchPreference secondMatchPreference) {
+    return (firstMatchPreference == MatchPreference.ANY)
+        ? secondMatchPreference
+        : firstMatchPreference;
+  }
```

Comment:
Since you refactored the methods to combine them, would this logic already be accounted for in `getCompatibleMatchPreference()`?

Reply:
oops yeah, forgot to delete this! thanks!